xnmt/translator.py

```diff
@@ -191,3 +194,190 @@ def html_report(self, context=None):
+    self.scale_emb = self.input_dim ** 0.5
+    self.initialize_position_encoding(500, input_dim)  # TODO: parametrize this
+
+  def initialize_generator(self, **kwargs):
+    if kwargs.get("len_norm_type", None) is None:
+      len_norm = xnmt.length_normalization.NoNormalization()
+    else:
+      len_norm = xnmt.serializer.YamlSerializer().initialize_object(kwargs["len_norm_type"])
+    search_args = {}
+    if kwargs.get("max_len", None) is not None:
+      search_args["max_len"] = kwargs["max_len"]
+      self.max_len = kwargs.get("max_len", 50)
+    if kwargs.get("beam", None) is None:
+      self.search_strategy = GreedySearch(**search_args)
+    else:
+      search_args["beam_size"] = kwargs.get("beam", 1)
+      search_args["len_norm"] = len_norm
+      # self.search_strategy = TransformerBeamSearch(**search_args)
+    self.report_path = kwargs.get("report_path", None)
+    self.report_type = kwargs.get("report_type", None)
+
+  def initialize_training_strategy(self, training_strategy):
+    self.loss_calculator = training_strategy
+
+  def set_reporting_src_vocab(self, src_vocab):
+    """
+    Sets source vocab for reporting purposes.
+    """
+    self.reporting_src_vocab = src_vocab
+
+  def make_attention_mask(self, source_block, target_block):
+    mask = (target_block[:, None, :] <= 0) * (source_block[:, :, None] <= 0)
+    # (batch, source_length, target_length)
+    return mask
+
+  def make_history_mask(self, block):
+    batch, length = block.shape
+    arange = np.arange(length)
+    history_mask = (arange[None,] <= arange[:, None])[None,]
+    history_mask = np.broadcast_to(history_mask, (batch, length, length))
+    return history_mask
+
+  def mask_embeddings(self, embeddings, mask):
+    """
+    We convert the embeddings of masked input sequence to zero vector
+    """
+    (embed_dim, _), _ = embeddings.dim()
+    temp_mask = np.repeat(1. - mask[:, None, :], embed_dim, axis=1)
+    temp_mask = dy.inputTensor(np.moveaxis(temp_mask, [1, 0, 2], [0, 2, 1]), batched=True)
+    embeddings = dy.cmult(embeddings, temp_mask)
+    return embeddings
+
+  def initialize_position_encoding(self, length, n_units):
+    # Implementation in the Google tensor2tensor repo
+    channels = n_units
+    position = np.arange(length, dtype='f')
+    num_timescales = channels // 2
+    log_timescale_increment = (np.log(10000. / 1.) / (float(num_timescales) - 1))
+    inv_timescales = 1. * np.exp(np.arange(num_timescales).astype('f') * -log_timescale_increment)
+    scaled_time = np.expand_dims(position, 1) * np.expand_dims(inv_timescales, 0)
+    signal = np.concatenate([np.sin(scaled_time), np.cos(scaled_time)], axis=1)
+    signal = np.reshape(signal, [1, length, channels])
+    self.position_encoding_block = np.transpose(signal, (0, 2, 1))
+
+  def make_input_embedding(self, emb_block, length):
+    emb_block = emb_block * self.scale_emb
+    emb_block += dy.inputTensor(self.position_encoding_block[0, :, :length])
```

Comment:
I think we should add a check here to see if length is greater than the length of `position_encoding_block`, and if so, re-call `initialize_positional_encoding(length*2, num_units)`

Reply:
(I had a crash because I had a sentence larger than 500)